src/muz/spacer/spacer_convex_closure.cpp

```diff
@@ -87,7 +88,84 @@ unsigned convex_closure::reduce_dim() {
     return m_dim - ker.num_rows();
 }
 
-// generates linear equalities implied by m_data
+// For row \p row in m_kernel, construct the equality:
+//
+// row * m_dim_vars = 0
+//
+// In the equality, exactly one variable from  m_dim_vars is on the lhs
+void convex_closure::generate_lin_deps_for_row(const vector<rational> &row,
+                                               expr_ref &res) {
+    // contains the right hand side of an equality
+    expr_ref_buffer rw(m);
+    // index of first non zero element in row
+    int pv = -1;
+    // are we constructing rhs or lhs
+    bool rhs = false;
+    // coefficient of m_dim_vars[pv]
+    rational coeff(1);
+    // the elements in row are the coefficients of m_dim_vars
+    // some elements should go to the rhs, in which case the signs are
+    // changed
+    for (unsigned j = 0; j < row.size(); j++) {
+        rational val = row.get(j);
+        SASSERT(val.is_int());
+        if (val.is_zero()) continue;
+        if (!rhs) {
+            // Cannot re-write the last element
+            if (j == row.size() - 1) continue;
+            SASSERT(pv == -1);
+            pv = j;
+            rhs = true;
+            // In integer echelon form, the pivot need not be 1
+            coeff = val;
+        } else {
+            expr_ref prod(m);
+            if (j != row.size() - 1) {
+                prod = m_dim_vars.get(j);
+                mul_by_rat(prod, -1 * val * m_lcm);
+            } else {
+                if (m_arith.is_int(m_dim_vars.get(pv))) {
+                    prod = m_arith.mk_int(-1 * val);
+                } else if (m_arith.is_real(m_dim_vars.get(pv))) {
+                    prod = m_arith.mk_real(-1 * val);
+                } else if (m_bv.is_bv(m_dim_vars.get(pv))) {
+                    prod = m_bv.mk_numeral(-1 * val, m_bv_sz);
+                }
+            }
+            SASSERT(prod.get());
+            rw.push_back(prod);
+        }
+    }
+
+    // make sure that there is a non-zero entry
+    SASSERT(pv != -1);
+
+    if (rw.size() == 0) {
+        expr_ref rhs(m);
+        if (m_arith.is_int(m_dim_vars.get(pv)))
+            rhs = m_arith.mk_int(rational::zero());
+        else if (m_arith.is_real(m_dim_vars.get(pv)))
+            rhs = m_arith.mk_real(rational::zero());
+        else if (m_bv.is_bv(m_dim_vars.get(pv)))
+            rhs = m_bv.mk_numeral(rational::zero(), m_bv_sz);
+        res = m_arith.mk_eq(m_dim_vars.get(pv), rhs);
+        return;
+    }
+
+    // TODO: n-ary bvadd is not in SMT-LIB standard. Is not supported in Z3
+    res = m_is_arith ? m_arith.mk_add(rw.size(), rw.c_ptr())
+        : m.mk_app(m_bv.get_fid(), OP_BADD, rw.size(),
```

Comment:
factor `mk_app` into `mk_bvadd` local method and shift comments about n-ary bvadd in there. It might be supported by Z3, just needs to be checked.